mmgen/models/architectures/sngan_proj/generator_discriminator.py

```diff
@@ -168,119 +182,115 @@ class SNGANDiscriminator(nn.Module):
-                        input_scale, [chn for chn in channel_cfg.keys()]))
+                raise KeyError(f'`input_scale={input_scale} is not found in '
+                               '`channel_cfg`, only support configs for '
+                               f'{[chn for chn in channel_cfg.keys()]}')
             self.channel_factor_list = channel_cfg[input_scale]
 
+        # get config for downsample coresponding to the input_scale
+        # downsample_cfg: whether apply downsample in this block
         downsample_cfg = self._defualt_downsample_list_cfg \
             if downsample_cfg is None else downsample_cfg
         if isinstance(downsample_cfg, dict):
             if input_scale not in downsample_cfg:
-                raise KeyError(
-                    '`output_scale={} is not found in `downsample_cfg`, '
-                    'only support configs for {}'.format(
-                        input_scale, [chn for chn in downsample_cfg.keys()]))
+                raise KeyError(f'`output_scale={input_scale} is not found in '
+                               '`downsample_cfg`, only support configs for '
+                               f'{[chn for chn in downsample_cfg.keys()]}')
             self.downsample_list = downsample_cfg[input_scale]
 
         if len(downsample_cfg) != len(channel_cfg):
-            raise ValueError('TODO')
+            raise ValueError(
+                '`downsample_cfg` should have same length with `channel_cfg`, '
+                f'but receive {len(downsample_cfg)} and {len(channel_cfg)}.')
 
-        self.from_rgb = SNGAN_DisHeadResBlock(input_channels, base_channels,
-                                              **self.head_cfg)
+        self.from_rgb = build_from_cfg(
+            self.from_rgb_cfg, MODULES,
+            dict(in_channels=input_channels, out_channels=base_channels))
 
-        self.res_block = nn.ModuleList()
+        self.blocks = nn.ModuleList()
         for idx in range(len(downsample_cfg)):
             factor_input = 1 if idx == 0 else self.channel_factor_list[idx - 1]
             factor_output = self.channel_factor_list[idx]
-            res_block_cfg_ = deepcopy(self.res_block_cfg)
-            res_block_cfg_['downsample'] = self.downsample_list[idx]
-            self.res_block.append(
-                SNGAN_DisResBlock(factor_input * base_channels,
-                                  factor_output * base_channels,
-                                  **res_block_cfg_))
 
-        self.decision = spectral_norm(
-            nn.Linear(factor_output * base_channels, 1, bias=False))
+            # get block-specific config
+            block_cfg_ = deepcopy(self.blocks_cfg)
+            block_cfg_['downsample'] = self.downsample_list[idx]
+            block_cfg_['in_channels'] = factor_input * base_channels
+            block_cfg_['out_channels'] = factor_output * base_channels
+            self.blocks.append(build_from_cfg(block_cfg_, MODULES))
```

Comment:
Please pass some necessary args in default args like `with_spectral_norm` and `act_cfg`.

Reply:
`with_spectral_norm` and `act_cfg` are passed into `self.blocks_cfg` in line 214-214.